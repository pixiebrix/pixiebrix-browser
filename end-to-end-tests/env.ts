/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { loadEnv } from "../scripts/env";

process.env.ENV_FILE ??= ".env.development";
loadEnv();

const requiredEnvVariables = [
  "SERVICE_URL",
  "E2E_TEST_USER_EMAIL_UNAFFILIATED",
  "E2E_TEST_USER_PASSWORD_UNAFFILIATED",
  "SHADOW_DOM",
] as const;

// It's not strictly required for the test run itself, but the extension manifest.json must have been built with
// "REQUIRE_OPTIONAL_PERMISSIONS_IN_MANIFEST" defined to avoid test failures due to missing permissions when activating certain mods.
const optionalEnvVariables = [
  "CI",
  "SLOWMO",
  "PWDEBUG",
  "REQUIRE_OPTIONAL_PERMISSIONS_IN_MANIFEST",
] as const;

type RequiredEnvVariables = Record<
  (typeof requiredEnvVariables)[number],
  string
>;
type OptionalEnvVariables = Record<
  (typeof optionalEnvVariables)[number],
  string | undefined
>;

export const assertRequiredEnvVariables = () => {
  for (const key of requiredEnvVariables) {
    // eslint-disable-next-line security/detect-object-injection -- key is a constant
    if (process.env[key] === undefined) {
      throw new Error(
        `Required environment variable is not configured: ${key}`,
      );
    }

    // eslint-disable-next-line security/detect-object-injection -- key is a constant
    if (typeof process.env[key] !== "string") {
      // For the time being we expect all of our requiredEnvVariables to be strings
      throw new TypeError(
        `Required environment variable is not configured: ${key}`,
      );
    }
  }
};

export const {
  SERVICE_URL,
  E2E_TEST_USER_EMAIL_UNAFFILIATED,
  E2E_TEST_USER_PASSWORD_UNAFFILIATED,
} = process.env as RequiredEnvVariables;

export const { CI, SLOWMO, PWDEBUG } = process.env as OptionalEnvVariables;
