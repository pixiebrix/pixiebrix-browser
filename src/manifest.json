{
  "manifest_version": 2,
  "author": "PixieBrix, Inc.",
  "name": "PixieBrix",
  "short_name": "PixieBrix",
  "description": "Customize, extend, and integrate your favorite SaaS apps without writing a line of code!",
  "homepage_url": "https://www.pixiebrix.com",
  "minimum_chrome_version": "95.0",
  "browser_specific_settings": {
    "gecko": {
      "id": "extension@pixiebrix.com",
      "strict_min_version": "91.0"
    }
  },
  "icons": {
    "16": "icons/logo16.png",
    "32": "icons/logo32.png",
    "48": "icons/logo48.png",
    "128": "icons/logo128.png"
  },
  "content_security_policy": "script-src 'self' https://apis.google.com 'unsafe-eval'; font-src 'self' https://fonts.gstatic.com; connect-src 'self' https:; object-src 'self'; img-src 'self' data: https:; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; frame-src 'self' https: https://*.googleapis.com https://docs.google.com",
  "content_scripts": [
    {
      "matches": ["https://*/*"],
      "js": ["contentScript.js"],
      "css": ["contentScript.css"],
      "run_at": "document_idle"
    },
    {
      "matches": ["https://*/*"],
      "js": ["loadActivationEnhancements.js"],
      "run_at": "document_end"
    }
  ],
  "sandbox": {
    "pages": ["sandbox.html"]
  },
  "optional_permissions": ["clipboardWrite"],
  "permissions": [
    "activeTab",
    "storage",
    "identity",
    "tabs",
    "webNavigation",
    "contextMenus",
    "*://*/*",
    "<all_urls>"
  ],
  "devtools_page": "devtools.html",
  "externally_connectable": {
    "matches": ["https://*.pixiebrix.com/*"]
  },
  "storage": {
    "managed_schema": "managedStorageSchema.json"
  },
  "web_accessible_resources": [
    "css/*",
    "bundles/*",
    "sandbox.html",
    "frame.html",
    "frame.css",
    "sidebar.html",
    "sidebar.css",
    "pageEditor.css",
    "pageScript.js",
    "ephemeralForm.html",
    "ephemeralPanel.html",
    "ephemeralModal.css",
    "audio/*",
    "user-icons/*",
    "img/*"
  ],
  "browser_action": {
    "default_title": "PixieBrix",
    "default_icon": {
      "16": "icons/inactive/logo16.png",
      "32": "icons/inactive/logo32.png",
      "48": "icons/inactive/logo48.png",
      "128": "icons/inactive/logo128.png"
    }
  },
  "commands": {
    "toggle-quick-bar": {
      "description": "Toggle Quick Bar",
      "suggested_key": {
        "default": "Ctrl+M",
        "mac": "Command+M"
      }
    }
  },
  "background": {
    "scripts": ["grayIconWhileLoading.js", "background.js"]
  },
  "options_ui": {
    "page": "options.html",
    "open_in_tab": true
  }
}
