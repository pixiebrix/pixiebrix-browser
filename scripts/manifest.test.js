/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { omit } from "lodash";
import manifest from "../applications/browser-extension/src/manifest.json";
import { loadEnv } from "./env.mjs";
import customizeManifest from "./manifest.mjs";

loadEnv();

const cleanCustomize = (...args) => omit(customizeManifest(...args), ["key"]);

describe("customizeManifest", () => {
  describe("release builds", () => {
    test("stable", () => {
      expect(
        cleanCustomize(manifest, {
          // eslint-disable-next-line camelcase -- auto-inserted
          env: { ...process.env, npm_package_version: "1.8.13" },
          isProduction: true,
          manifestVersion: 3,
        }),
      ).toMatchSnapshot();
    });

    test("beta", () => {
      expect(
        cleanCustomize(manifest, {
          // eslint-disable-next-line camelcase -- auto-inserted
          env: { ...process.env, npm_package_version: "1.8.13" },
          isProduction: true,
          manifestVersion: 3,
          isBeta: true,
        }),
      ).toMatchSnapshot();
    });
  });

  describe("four digit versioning", () => {
    test("alpha version", () => {
      expect(
        cleanCustomize(manifest, {
          // eslint-disable-next-line camelcase -- auto-inserted
          env: { ...process.env, npm_package_version: "1.8.13-alpha.123" },
          isProduction: true,
          manifestVersion: 3,
        }),
      ).toContainEntry(["version", "1.8.13.1123"]);
    });

    test("beta version", () => {
      expect(
        cleanCustomize(manifest, {
          // eslint-disable-next-line camelcase -- auto-inserted
          env: { ...process.env, npm_package_version: "1.8.13-beta.123" },
          isProduction: true,
          manifestVersion: 3,
        }),
      ).toContainEntry(["version", "1.8.13.2123"]);
    });

    test("release version", () => {
      expect(
        cleanCustomize(manifest, {
          // eslint-disable-next-line camelcase -- auto-inserted
          env: { ...process.env, npm_package_version: "1.8.13" },
          isProduction: true,
          manifestVersion: 3,
        }),
      ).toContainEntry(["version", "1.8.13.3000"]);
    });
  });
});
