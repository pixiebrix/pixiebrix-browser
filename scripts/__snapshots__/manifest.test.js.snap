// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customizeManifest mv2 1`] = `
{
  "author": "PixieBrix, Inc.",
  "background": {
    "scripts": [
      "backgroundCanary.js",
      "background.js",
      "webext-storage-cache.js",
    ],
  },
  "browser_action": {
    "default_icon": {
      "128": "icons/inactive/logo128.png",
      "16": "icons/inactive/logo16.png",
      "32": "icons/inactive/logo32.png",
      "48": "icons/inactive/logo48.png",
    },
    "default_title": "PixieBrix",
  },
  "browser_specific_settings": {
    "gecko": {
      "id": "extension@pixiebrix.com",
      "strict_min_version": "91.0",
    },
  },
  "commands": {
    "toggle-quick-bar": {
      "description": "Toggle Quick Bar",
      "suggested_key": {
        "default": "Ctrl+M",
        "mac": "Command+M",
      },
    },
  },
  "content_scripts": [
    {
      "all_frames": true,
      "css": [
        "contentScript.css",
      ],
      "exclude_matches": [
        "https://*.googleapis.com/*",
      ],
      "js": [
        "contentScript.js",
      ],
      "match_about_blank": true,
      "matches": [
        "*://*/*",
      ],
      "run_at": "document_idle",
    },
    {
      "js": [
        "loadActivationEnhancements.js",
      ],
      "matches": [
        "https://*/*",
      ],
      "run_at": "document_end",
    },
    {
      "js": [
        "setExtensionIdInApp.js",
      ],
      "matches": [
        "https://*.pixiebrix.com/*",
        "http://127.0.0.1:8000/*",
        "http://127.0.0.1/*",
        "http://localhost/*",
        "http://localhost:8000/*",
      ],
      "run_at": "document_end",
    },
  ],
  "content_security_policy": "script-src 'self' 'unsafe-eval' http://localhost:8097; font-src 'self' https://fonts.gstatic.com; connect-src 'self' http: https: ws://localhost:8097/; object-src 'self'; img-src 'self' data: https: https://pixiebrix-marketplace-dev.s3.amazonaws.com; style-src 'self' 'unsafe-inline' https:; frame-src 'self' https:;",
  "description": "Customize, extend, and integrate your favorite SaaS apps without writing a line of code!",
  "devtools_page": "devtools.html",
  "externally_connectable": {
    "matches": [
      "https://*.pixiebrix.com/*",
      "http://127.0.0.1:8000/*",
      "http://127.0.0.1/*",
      "http://localhost/*",
      "http://localhost:8000/*",
    ],
  },
  "homepage_url": "https://www.pixiebrix.com",
  "icons": {
    "128": "icons/logo128.png",
    "16": "icons/logo16.png",
    "32": "icons/logo32.png",
    "48": "icons/logo48.png",
  },
  "manifest_version": 2,
  "minimum_chrome_version": "95.0",
  "name": "PixieBrix - Development",
  "optional_permissions": [
    "clipboardWrite",
  ],
  "options_ui": {
    "open_in_tab": true,
    "page": "options.html",
  },
  "permissions": [
    "activeTab",
    "storage",
    "identity",
    "tabs",
    "webNavigation",
    "contextMenus",
    "<all_urls>",
    "alarms",
  ],
  "sandbox": {
    "pages": [
      "sandbox.html",
    ],
  },
  "short_name": "PixieBrix",
  "storage": {
    "managed_schema": "managedStorageSchema.json",
  },
  "web_accessible_resources": [
    "css/*",
    "bundles/*",
    "sandbox.html",
    "frame.html",
    "frame.css",
    "sidebar.html",
    "sidebar.css",
    "pageEditor.css",
    "pageScript.js",
    "ephemeralForm.html",
    "walkthroughModal.html",
    "ephemeralPanel.html",
    "ephemeralModal.css",
    "DocumentView.css",
    "audio/*",
    "user-icons/*",
    "img/*",
  ],
}
`;

exports[`customizeManifest mv3 1`] = `
{
  "action": {
    "default_icon": {
      "128": "icons/inactive/logo128.png",
      "16": "icons/inactive/logo16.png",
      "32": "icons/inactive/logo32.png",
      "48": "icons/inactive/logo48.png",
    },
    "default_title": "PixieBrix",
  },
  "author": "PixieBrix, Inc.",
  "background": {
    "service_worker": "background.worker.js",
    "type": "module",
  },
  "browser_specific_settings": {
    "gecko": {
      "id": "extension@pixiebrix.com",
      "strict_min_version": "91.0",
    },
  },
  "commands": {
    "toggle-quick-bar": {
      "description": "Toggle Quick Bar",
      "suggested_key": {
        "default": "Ctrl+M",
        "mac": "Command+M",
      },
    },
  },
  "content_scripts": [
    {
      "all_frames": true,
      "css": [
        "contentScript.css",
      ],
      "exclude_matches": [
        "https://*.googleapis.com/*",
      ],
      "js": [
        "contentScript.js",
      ],
      "match_about_blank": true,
      "matches": [
        "*://*/*",
      ],
      "run_at": "document_idle",
    },
    {
      "js": [
        "loadActivationEnhancements.js",
      ],
      "matches": [
        "https://*/*",
      ],
      "run_at": "document_end",
    },
    {
      "js": [
        "setExtensionIdInApp.js",
      ],
      "matches": [
        "https://*.pixiebrix.com/*",
        "http://127.0.0.1:8000/*",
        "http://127.0.0.1/*",
        "http://localhost/*",
        "http://localhost:8000/*",
      ],
      "run_at": "document_end",
    },
  ],
  "content_security_policy": {
    "extension_pages": "script-src 'self' http://localhost:8097; font-src 'self' https://fonts.gstatic.com; connect-src 'self' http: https: ws://localhost:8097/; object-src 'self'; img-src 'self' data: https: https://pixiebrix-marketplace-dev.s3.amazonaws.com; style-src 'self' 'unsafe-inline' https:; frame-src 'self' https:;",
    "sandbox": "sandbox allow-scripts allow-forms; script-src 'self' 'unsafe-inline' 'unsafe-eval'; child-src 'self';",
  },
  "description": "Customize, extend, and integrate your favorite SaaS apps without writing a line of code!",
  "devtools_page": "devtools.html",
  "externally_connectable": {
    "matches": [
      "https://*.pixiebrix.com/*",
      "http://127.0.0.1:8000/*",
      "http://127.0.0.1/*",
      "http://localhost/*",
      "http://localhost:8000/*",
    ],
  },
  "homepage_url": "https://www.pixiebrix.com",
  "host_permissions": [
    "<all_urls>",
  ],
  "icons": {
    "128": "icons/logo128.png",
    "16": "icons/logo16.png",
    "32": "icons/logo32.png",
    "48": "icons/logo48.png",
  },
  "manifest_version": 3,
  "minimum_chrome_version": "116.0",
  "name": "PixieBrix - Development",
  "optional_permissions": [
    "clipboardWrite",
  ],
  "options_ui": {
    "open_in_tab": true,
    "page": "options.html",
  },
  "permissions": [
    "activeTab",
    "storage",
    "identity",
    "tabs",
    "webNavigation",
    "contextMenus",
    "alarms",
    "devtools",
    "scripting",
    "sidePanel",
  ],
  "sandbox": {
    "pages": [
      "sandbox.html",
    ],
  },
  "short_name": "PixieBrix",
  "side_panel": {
    "default_path": "sidebar.html",
  },
  "storage": {
    "managed_schema": "managedStorageSchema.json",
  },
  "web_accessible_resources": [
    {
      "matches": [
        "*://*/*",
      ],
      "resources": [
        "css/*",
        "bundles/*",
        "sandbox.html",
        "frame.html",
        "frame.css",
        "sidebar.html",
        "sidebar.css",
        "pageEditor.css",
        "pageScript.js",
        "ephemeralForm.html",
        "walkthroughModal.html",
        "ephemeralPanel.html",
        "ephemeralModal.css",
        "DocumentView.css",
        "audio/*",
        "user-icons/*",
        "img/*",
      ],
    },
  ],
}
`;
